packages/use-dataloader/src/useDataLoader/index.js

```diff
@@ -0,0 +1,145 @@
+    data: initialData,
+    error: undefined,
+  })
+
+  const keyRef = useRef()
+  const handlerRef = useRef()
+  const methodRef = useRef(method)
+  const onSuccessRef = useRef(onSuccess)
+  const onErrorRef = useRef(onError)
+
+  const isLoading = useMemo(() => status === StatusEnum.LOADING, [status])
+  const isIdle = useMemo(() => status === StatusEnum.IDLE, [status])
+  const isSuccess = useMemo(() => status === StatusEnum.SUCCESS, [status])
+  const isError = useMemo(() => status === StatusEnum.ERROR, [status])
+  const isPolling = useMemo(
+    () => handlerRef.current && (isSuccess || isLoading),
+    [isSuccess, isLoading],
+  )
+
+  const updateData = useCallback(
+    (newData, updateCache = true) => {
+      dispatch(Actions.createOnUpdateData(newData))
+      if (updateCache && key && newData) {
+        addCachedData(key, newData)
+      }
+    },
+    [addCachedData, key],
+  )
+
+  const handleRequest = useCallback(
+    async (cacheKey, args, force = false) => {
+      if (cacheKey && !data && getCachedData(cacheKey)) {
+        dispatch(Actions.createOnUpdateData(getCachedData(cacheKey)))
+      }
+      if (status === StatusEnum.LOADING && !force) {
+        return
+      }
+      try {
+        dispatch(Actions.createOnLoading())
+        const result = await methodRef.current?.(args)
+
+        if (result && cacheKey) addCachedData(cacheKey, result)
+
+        dispatch(Actions.createOnSuccess(result))
+
+        await onSuccessRef.current?.(result)
+      } catch (err) {
+        dispatch(Actions.createOnError(err))
+        await onErrorRef.current?.(err)
+      }
+    },
+    [addCachedData, data, getCachedData, status],
+  )
+
+  const launchRequest = useCallback(
+    (args, forceReload) => {
+      handleRequest(key, args, forceReload)
+      if (pollingInterval) {
+        if (handlerRef.current) {
+          clearInterval(handlerRef.current)
+        }
+
+        handlerRef.current = setInterval(
+          () => handleRequest(key, args),
+          pollingInterval,
+        )
+      }
+      if (key && typeof key === 'string') {
+        clearReload(key)
+        addReload(key, reloadArgs => handleRequest(key, reloadArgs))
+      }
+    },
+    [addReload, handleRequest, clearReload, key, pollingInterval],
+  )
+
+  useEffect(() => {
+    if (enabled && status === StatusEnum.IDLE) launchRequest()
+    // Why can't put empty array for componentDidMount, componentWillUnmount ??? No array act like componentDidMount and componentDidUpdate
+  }, [enabled, launchRequest, status])
+
+  useEffect(() => {
+    if (enabled && reloadOnKeyChange && key !== keyRef.current) {
+      keyRef.current = key
+      launchRequest()
+    }
+  }, [enabled, launchRequest, status, key, reloadOnKeyChange, initialData])
+
+  useLayoutEffect(() => {
+    methodRef.current = method
+  }, [method])
+
+  return {
+    state: status,
+    isLoading,
+    isIdle,
+    isSuccess,
+    isError,
+    isPolling,
+    data,
+    error,
+    updateData,
```

Comment:
This means the user will be able to manually update the result of a cached query from outside the data loader scope ? 
Shouldn't this function be kept private ?

Reply:
The user can change data manually if he want to edit them (to use it in a form or something else)

Reply:
For me it's a bad practice to let the user transform de data.

Imagine that you modify the date before a reload and you make put/patch/post request that goes wrong ? You will need to clear the data in this case ?
if you forget to clear the data you will show wrong information to your users ?

An other point :
you believe that the api will return everything you put in updateData but if there an error again and the data doesn't return what you think she will return. it's will not break at first but if you comme back later on this page and you didn't handle the use case where for exemple de data is a number and not a string ? or a boolean and not a number this will break your app